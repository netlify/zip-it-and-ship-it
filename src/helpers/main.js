/* eslint-disable node/no-unpublished-require */
const { join } = require('path')
const { promisify } = require('util')

const AdmZip = require('adm-zip')
const { dir: getTmpDir } = require('tmp-promise')

const { zipFunctions } = require('../..')

const FIXTURES_DIR = join(__dirname, '..', 'fixtures')

const zipNode = async function(t, fixture, length, opts, fixtureDir) {
  const { files, tmpDir } = await zipFixture(t, fixture, length, opts, fixtureDir)
  await requireExtractedFiles(t, files)
  return { files, tmpDir }
}

const zipFixture = async function(t, fixture, length, opts, fixtureDir) {
  const { path: tmpDir } = await getTmpDir({ prefix: 'zip-it-test' })
  const { files } = await zipCheckFunctions(t, fixture, tmpDir, length, opts, fixtureDir)
  return { files, tmpDir }
}

const zipCheckFunctions = async function(t, fixture, tmpDir, length = 1, opts, fixtureDir = FIXTURES_DIR) {
  const files = await zipFunctions(`${fixtureDir}/${fixture}`, tmpDir, opts)

  t.true(Array.isArray(files))
  t.is(files.length, length)

  return { files, tmpDir }
}

const requireExtractedFiles = async function(t, files) {
  await unzipFiles(files)

  const jsFiles = files.map(replaceUnzipPath).map(require)
  t.true(jsFiles.every(Boolean))
}

const unzipFiles = async function(files) {
  await Promise.all(files.map(unzipFile))
}

const unzipFile = async function({ path }) {
  const zip = new AdmZip(path)
  const pExtractAll = promisify(zip.extractAllToAsync.bind(zip))
  await pExtractAll(`${path}/..`, false)
}

const replaceUnzipPath = function({ path }) {
  return path.replace('.zip', '.js')
}

module.exports = {
  zipNode,
  zipFixture,
  unzipFiles,
  zipCheckFunctions,
  FIXTURES_DIR
}
