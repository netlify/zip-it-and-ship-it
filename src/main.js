const { readdir, lstat } = require('fs')
const { join, resolve, dirname, basename, extname } = require('path')

const cpFile = require('cp-file')
const locatePath = require('locate-path')
const makeDir = require('make-dir')
const pMap = require('p-map')
const promisify = require('util.promisify')

const { listNodeFiles } = require('./dependencies')
const { isGoExe, zipGoExe } = require('./go')
const { zipNodeJs } = require('./node')

const pReaddir = promisify(readdir)
const pLstat = promisify(lstat)

// Zip `srcFolder/*` (Node.js or Go files) to `destFolder/*.zip` so it can be
// used by AWS Lambda
// TODO: remove `skipGo` option in next major release
const zipFunctions = async function(srcFolder, destFolder, { parallelLimit = 5, skipGo, zipGo } = {}) {
  const srcPaths = await getSrcPaths(srcFolder)

  const zipped = await pMap(srcPaths, srcPath => zipFunction(srcPath, destFolder, { skipGo, zipGo }), {
    concurrency: parallelLimit
  })
  return zipped.filter(Boolean)
}

const zipFunction = async function(srcPath, destFolder, { skipGo = true, zipGo = !skipGo } = {}) {
  const { runtime, filename, extension, stat, mainFile, srcFiles } = await getFunctionInfo(srcPath)

  if (runtime === undefined) {
    return
  }

  await makeDir(destFolder)

  if (runtime === 'js') {
    if (extension === '.zip') {
      const destPath = join(destFolder, filename)
      await cpFile(srcPath, destPath)
      return { path: destPath, runtime }
    } else {
      const destPath = join(destFolder, `${basename(filename, '.js')}.zip`)
      await zipNodeJs(srcFiles, destPath, filename, mainFile)
      return { path: destPath, runtime }
    }
  }

  if (runtime === 'go') {
    if (zipGo) {
      const destPath = join(destFolder, `${filename}.zip`)
      await zipGoExe(srcPath, destPath, filename, stat)
      return { path: destPath, runtime }
    } else {
      const destPath = join(destFolder, filename)
      await cpFile(srcPath, destPath)
      return { path: destPath, runtime }
    }
  }
}

// List all Netlify Functions for a specific directory
const listFunctions = async function(srcFolder) {
  const srcPaths = await getSrcPaths(srcFolder)
  const functionInfos = await Promise.all(srcPaths.map(getFunctionInfo))
  return functionInfos.filter(hasMainFile).map(getListedFunction)
}

const getSrcPaths = async function(srcFolder) {
  const filenames = await listFilenames(srcFolder)
  const srcPaths = filenames.map(filename => resolve(srcFolder, filename))
  return srcPaths
}

const listFilenames = async function(srcFolder) {
  try {
    return await pReaddir(srcFolder)
  } catch (error) {
    throw new Error(`Functions folder does not exist: ${srcFolder}`)
  }
}

const getFunctionInfo = async function(srcPath) {
  const { filename, stat, mainFile, extension, srcDir } = await getSrcInfo(srcPath)

  if (mainFile === undefined) {
    return {}
  }

  if (extension === '.zip') {
    return { runtime: 'js', filename, stat, mainFile, extension, srcFiles: [srcPath] }
  }

  if (extension === '.js') {
    const srcFiles = await listNodeFiles(srcPath, filename, mainFile, srcDir, stat)
    return { runtime: 'js', filename, stat, mainFile, extension, srcFiles }
  }

  if (await isGoExe(srcPath)) {
    return { runtime: 'go', filename, stat, mainFile, extension, srcFiles: [srcPath] }
  }

  return {}
}

const getSrcInfo = async function(srcPath) {
  const filename = basename(srcPath)
  if (filename === 'node_modules') {
    return {}
  }

  const stat = await pLstat(srcPath)
  const mainFile = await getMainFile(srcPath, filename, stat)
  if (mainFile === undefined) {
    return {}
  }

  const extension = extname(mainFile)
  const srcDir = stat.isDirectory() ? srcPath : dirname(srcPath)
  return { filename, stat, mainFile, extension, srcDir }
}

// Each `srcPath` can also be a directory with an `index.js` file or a file
// using the same filename as its directory
const getMainFile = function(srcPath, filename, stat) {
  if (!stat.isDirectory()) {
    return srcPath
  }

  return locatePath([join(srcPath, `${filename}.js`), join(srcPath, 'index.js')], { type: 'file' })
}

const hasMainFile = function({ mainFile }) {
  return mainFile !== undefined
}

const getListedFunction = function({ mainFile, srcFiles, runtime, extension }) {
  return { mainFile, srcFiles, runtime, extension }
}

module.exports = { zipFunctions, zipFunction, listFunctions }
