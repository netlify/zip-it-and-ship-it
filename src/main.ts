import { extname } from 'path'

import { Config } from './config.js'
import { FeatureFlags, getFlags } from './feature_flags.js'
import { FunctionSource } from './function.js'
import { getFunctionFromPath, getFunctionsFromPaths } from './runtimes/index.js'
import { findISCDeclarationsInPath, ISCValues } from './runtimes/node/in_source_config/index.js'
import { GetSrcFilesFunction, RuntimeName } from './runtimes/runtime.js'
import { listFunctionsDirectories, resolveFunctionsDirectories } from './utils/fs.js'

export { zipFunction, zipFunctions } from './zip.js'

interface ListedFunction {
  name: string
  mainFile: string
  runtime: RuntimeName
  extension: string
  schedule?: string
}

type ListedFunctionFile = ListedFunction & {
  srcFile: string
}

interface ListFunctionsOptions {
  basePath?: string
  config?: Config
  featureFlags?: FeatureFlags
  parseISC?: boolean
}

interface AugmentedFunctionSource extends FunctionSource {
  inSourceConfig?: ISCValues
}

const augmentWithISC = async (func: FunctionSource): Promise<AugmentedFunctionSource> => {
  // ISC is currently only supported in JavaScript and TypeScript functions.
  if (func.runtime.name !== 'js') {
    return func
  }

  const inSourceConfig = await findISCDeclarationsInPath(func.mainFile)

  return { ...func, inSourceConfig }
}

// List all Netlify Functions main entry files for a specific directory
export const listFunctions = async function (
  relativeSrcFolders: string | string[],
  {
    featureFlags: inputFeatureFlags,
    config,
    parseISC = false,
  }: { featureFlags?: FeatureFlags; config?: Config; parseISC?: boolean } = {},
) {
  const featureFlags = getFlags(inputFeatureFlags)
  const srcFolders = resolveFunctionsDirectories(relativeSrcFolders)
  const paths = await listFunctionsDirectories(srcFolders)
  const functionsMap = await getFunctionsFromPaths(paths, { featureFlags, config })
  const functions = [...functionsMap.values()]
  const augmentedFunctions = parseISC ? await Promise.all(functions.map(augmentWithISC)) : functions

  return augmentedFunctions.map(getListedFunction)
}

// Finds a function at a specific path.
export const listFunction = async function (
  path: string,
  {
    featureFlags: inputFeatureFlags,
    config,
    parseISC = false,
  }: { featureFlags?: FeatureFlags; config?: Config; parseISC?: boolean } = {},
) {
  const featureFlags = getFlags(inputFeatureFlags)
  const func = await getFunctionFromPath(path, { featureFlags, config })

  if (!func) {
    return
  }

  const augmentedFunction = parseISC ? await augmentWithISC(func) : func

  return getListedFunction(augmentedFunction)
}

// List all Netlify Functions files for a specific directory
export const listFunctionsFiles = async function (
  relativeSrcFolders: string | string[],
  { basePath, config, featureFlags: inputFeatureFlags, parseISC = false }: ListFunctionsOptions = {},
) {
  const featureFlags = getFlags(inputFeatureFlags)
  const srcFolders = resolveFunctionsDirectories(relativeSrcFolders)
  const paths = await listFunctionsDirectories(srcFolders)
  const functionsMap = await getFunctionsFromPaths(paths, { config, featureFlags })
  const functions = [...functionsMap.values()]
  const augmentedFunctions = parseISC ? await Promise.all(functions.map(augmentWithISC)) : functions
  const listedFunctionsFiles = await Promise.all(
    augmentedFunctions.map((func) => getListedFunctionFiles(func, { basePath, featureFlags })),
  )

  return listedFunctionsFiles.flat()
}

const getListedFunction = function ({
  runtime,
  name,
  mainFile,
  extension,
  config,
  inSourceConfig,
}: AugmentedFunctionSource): ListedFunction {
  return { name, mainFile, runtime: runtime.name, extension, schedule: inSourceConfig?.schedule ?? config.schedule }
}

const getListedFunctionFiles = async function (
  func: AugmentedFunctionSource,
  options: { basePath?: string; featureFlags: FeatureFlags },
): Promise<ListedFunctionFile[]> {
  const srcFiles = await getSrcFiles({ ...func, ...options })

  return srcFiles.map((srcFile) => ({ ...getListedFunction(func), srcFile, extension: extname(srcFile) }))
}

const getSrcFiles: GetSrcFilesFunction = async function ({ extension, runtime, srcPath, ...args }) {
  const { getSrcFiles: getRuntimeSrcFiles } = runtime

  if (extension === '.zip' || typeof getRuntimeSrcFiles !== 'function') {
    return [srcPath]
  }

  return await getRuntimeSrcFiles({
    extension,
    runtime,
    srcPath,
    ...args,
  })
}
