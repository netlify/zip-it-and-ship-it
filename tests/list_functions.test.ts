import { join } from 'path'

import { describe, expect, test } from 'vitest'

import { listFunctions } from '../src/main.js'

import { FIXTURES_DIR, normalizeFiles } from './helpers/main.js'

describe('listFunctions', () => {
  test('Can list function main files with listFunctions()', async () => {
    const fixtureDir = `${FIXTURES_DIR}/list`
    const functions = await listFunctions(fixtureDir)
    expect(functions).toEqual(
      [
        { schedule: undefined, name: 'test', mainFile: 'test.zip', runtime: 'js', extension: '.zip' },
        { schedule: undefined, name: 'test', mainFile: 'test.js', runtime: 'js', extension: '.js' },
        { schedule: undefined, name: 'five', mainFile: 'five/index.ts', runtime: 'js', extension: '.ts' },
        { schedule: undefined, name: 'four', mainFile: 'four.js/four.js.js', runtime: 'js', extension: '.js' },
        { schedule: undefined, name: 'one', mainFile: 'one/index.js', runtime: 'js', extension: '.js' },
        { schedule: undefined, name: 'two', mainFile: 'two/two.js', runtime: 'js', extension: '.js' },
        { schedule: undefined, name: 'test', mainFile: 'test', runtime: 'go', extension: '' },
      ].map(normalizeFiles.bind(null, fixtureDir)),
    )
  })

  test('Can list function main files from multiple source directories with listFunctions()', async () => {
    const fixtureDir = `${FIXTURES_DIR}/multiple-src-directories`
    const functions = await listFunctions([
      join(fixtureDir, '.netlify', 'internal-functions'),
      join(fixtureDir, 'netlify', 'functions'),
    ])

    expect(functions).toEqual(
      [
        {
          schedule: undefined,
          name: 'function',
          mainFile: '.netlify/internal-functions/function.js',
          runtime: 'js',
          extension: '.js',
        },
        {
          schedule: undefined,
          name: 'function_internal',
          mainFile: '.netlify/internal-functions/function_internal.js',
          runtime: 'js',
          extension: '.js',
        },
        {
          schedule: undefined,
          name: 'function',
          mainFile: 'netlify/functions/function.js',
          runtime: 'js',
          extension: '.js',
        },
        {
          schedule: undefined,
          name: 'function_user',
          mainFile: 'netlify/functions/function_user.js',
          runtime: 'js',
          extension: '.js',
        },
      ].map(normalizeFiles.bind(null, fixtureDir)),
    )
  })

  test('listFunctions surfaces schedule config property', async () => {
    const functions = await listFunctions(join(FIXTURES_DIR, 'many-functions'), {
      config: {
        five: {
          schedule: '@daily',
        },
      },
    })
    const five = functions.find((func) => func.name === 'five')
    expect(five?.schedule).toBe('@daily')
  })

  test('listFunctions includes in-source config declarations', async () => {
    const functions = await listFunctions(join(FIXTURES_DIR, 'in-source-config', 'functions'), {
      parseISC: true,
    })
    const FUNCTIONS_COUNT = 13
    expect(functions.length).toBe(FUNCTIONS_COUNT)
    functions.forEach((func) => {
      expect(func.schedule).toBe('@daily')
    })
  })
})
