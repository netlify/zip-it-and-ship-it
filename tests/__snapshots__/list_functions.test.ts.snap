// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`listFunctions > V2 API > listFunctions includes runtimeAPIVersion when parseISC is true 1`] = `
{
  "displayName": undefined,
  "extension": ".ts",
  "generator": undefined,
  "inputModuleFormat": "esm",
  "mainFile": "function.ts",
  "name": "function",
  "runtime": "js",
  "runtimeAPIVersion": 2,
  "schedule": undefined,
  "srcFile": undefined,
  "tsConfig": undefined,
}
`;

exports[`listFunctions > v1 > Can list function main files from multiple source directories with listFunctions() 1`] = `
[
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": ".netlify/internal-functions/function.js",
    "name": "function",
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": ".netlify/internal-functions/function_internal.js",
    "name": "function_internal",
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "netlify/functions/function.js",
    "name": "function",
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "netlify/functions/function_user.js",
    "name": "function_user",
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcFile": undefined,
    "tsConfig": undefined,
  },
]
`;

exports[`listFunctions > v1 > Can list function main files with listFunctions() 1`] = `
[
  {
    "displayName": undefined,
    "extension": ".zip",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "test.zip",
    "name": "test",
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "test.js",
    "name": "test",
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".ts",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "five/index.ts",
    "name": "five",
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "four.js/four.js.js",
    "name": "four",
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "one/index.js",
    "name": "one",
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "two/two.js",
    "name": "two",
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": "",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "test",
    "name": "test",
    "runtime": "go",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcFile": undefined,
    "tsConfig": undefined,
  },
]
`;

exports[`listFunctions > v1 > listFunctions includes in-source config declarations 1`] = `
[
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "cjs",
    "mainFile": "cron_cjs.js",
    "name": "cron_cjs",
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "cjs",
    "mainFile": "cron_cjs_exports.js",
    "name": "cron_cjs_exports",
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "cjs",
    "mainFile": "cron_cjs_renamed.js",
    "name": "cron_cjs_renamed",
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm.js",
    "name": "cron_esm",
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_additional_schedule_import.js",
    "name": "cron_esm_additional_schedule_import",
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_no_direct_export.js",
    "name": "cron_esm_no_direct_export",
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_no_direct_export_renamed.js",
    "name": "cron_esm_no_direct_export_renamed",
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_no_direct_export_renamed_reassigned.js",
    "name": "cron_esm_no_direct_export_renamed_reassigned",
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_renamed.js",
    "name": "cron_esm_renamed",
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_schedule_multi_import.js",
    "name": "cron_esm_schedule_multi_import",
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_schedule_reassigned.js",
    "name": "cron_esm_schedule_reassigned",
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".ts",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_ts.ts",
    "name": "cron_ts",
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcFile": undefined,
    "tsConfig": undefined,
  },
  {
    "displayName": undefined,
    "extension": ".ts",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_ts_renamed.ts",
    "name": "cron_ts_renamed",
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcFile": undefined,
    "tsConfig": undefined,
  },
]
`;
